### DISCUSSION TITLE
Driving the LEDs with a kernel module

### DISCUSSION THREAD

**Discussion by kirkster96** [zangman/de10-nano#19] [General]
I am currently working on trying to recreate [Step 10 - Driving the LEDs with a kernel module](https://bitlog.it/20170820_building_embedded_linux_for_the_terasic_de10-nano.html#:~:text=Step%2010%20%2D%20Driving%20the%20LEDs%20with%20a%20kernel%20module), but have it work specifically with the setup for this repo.

I have gone through [Configuring the Device Tree](https://github.com/zangman/de10-nano/blob/master/docs/Configuring-the-Device-Tree.md) which allows the mmap() syscall to work for controlling the LEDs in a user program. This made sense to me and worked for me.

I have been studying  [Step 10 - Driving the LEDs with a kernel module](https://bitlog.it/20170820_building_embedded_linux_for_the_terasic_de10-nano.html#:~:text=Step%2010%20%2D%20Driving%20the%20LEDs%20with%20a%20kernel%20module) and [12. Writing a Simple Driver on rocketboards](https://rocketboards.org/foswiki/Documentation/EmbeddedLinuxBeginnerSGuide#11).

I used the GHRD and generated soc_system.dts, but if I try to use this device tree unmodified there seems to be some compatibility issues (for example, packets are dropped when I ping the DE10-nano over the home network).

I noticed that in my soc_system.dts there is this:
```
custom_leds_0: leds@0x100000000 {
				compatible = "dsa,custom_leds-1.0", "dev,custom-leds";
				reg = <0x00000001 0x00000000 0x00000008>;
				clocks = <&clk_0>;
			}; //end leds@0x100000000 (custom_leds_0)
```
I tried to add this inside of linux-socfpga/arch/arm/boot/dts/socfpga.dtsi inside of the soc {...} block. I modified it to look like this:

```
		custom_leds_0: leds@0xff400000 {
				compatible = "dsa,custom_leds-1.0", "dev,custom-leds";
				reg = <0xff400000 0x000008>;
				clocks = <&l4_main_clk>;
			}; //end leds@0xff400000 (custom_leds_0)
```

The .dtb builds successfully, and I thought my changes would make the register point to where the lightweight HPS-FPGA bridge is located (changes from [Configuring the Device Tree](https://github.com/zangman/de10-nano/blob/master/docs/Configuring-the-Device-Tree.md) are also added and FPGA bridges are all enabled in this custom device tree). 

After changing the device tree and rebooting, my IP show up under /sys/device/platform/soc/ and my LKM successfully runs the probe function. My LKM is exactly the  [12. Writing a Simple Driver on rocketboards](https://rocketboards.org/foswiki/Documentation/EmbeddedLinuxBeginnerSGuide#11), with small necessary changes to get it to build successfully. However, the LEDs do not respond to the driver when it tries to turn them on. This must mean that on entry, the probe function is not writting to the correct memory location where the lw bridge is located.

I am going to continue studying this, but if someone has more expience with correctly adding custom IPs to the device tree and can advise on this please let me know! If I get it working, I am planning on writing a PR to expand this guide with that new step.

Best regards,

Cameron Kirk

**Comment by kirkster96** [comment]
Ah, really quick update. I just got the LKM LED driver to work successfully. I was able to find this guide: [HOWTO Create a Device Tree](https://rocketboards.org/foswiki/Documentation/HOWTOCreateADeviceTree) on rocketboards and it gave me exactly what I was missing.

It's section on using sopc-create-header-files gave me the correct base address I was looking for. I updated my change to socfpga.dtsi with this:

```
custom_leds_0: leds@0xff200000 {
				compatible = "dsa,custom_leds-1.0", "dev,custom-leds";
				reg = <0xff200000 0x000008>;
				clocks = <&l4_main_clk>;
		}; //end leds@0xff200000 (custom_leds_0)
```

I will work on creating a PR containing instructions for how to do this.

**Reply by kirkster96** [reply]
Wow! It has just come to my attention that the [HOWTO Create a Device Tree](https://rocketboards.org/foswiki/Documentation/HOWTOCreateADeviceTree) is already in the References section of [Designing and Flashing the design](https://github.com/zangman/de10-nano/blob/master/docs/Building-SoC-Design.md). How did I miss this! 👌

**Comment by kirkster96** [comment]
The PR is now on this repo and I have created a video walkthrough of getting a linux driver to work for this repo's setup. Feel free to check it out now if you are interested:
[kirkster96/de10-nano/docs/Write-Linux-Driver.md](https://github.com/kirkster96/de10-nano/blob/master/docs/Write-Linux-Driver.md)
[YouTube - Write Linux Driver walkthrough](https://youtu.be/t3zkLh9-lUk)

**Comment by electrohike** [comment]
Quite a long time after, I'm reacting to this discussion because I'm stuck with an issue around the same topic : Right now, I can't build my own kernel module. I'm stuck when I try to run the make command to build my driver. The system answers that '/lib/modules/5.12.0zImage/build' does not exist.
That's true. The directory does not exist. Something is missing. From [that page](https://docs.kernel.org/kbuild/modules.html), I've understood that I need to install "the configuration and header files used in the (kernel) build". But were can I find and install that ?

I'm working on a DE10-Nano with the Debian image released [here](https://github.com/zangman/de10-nano/releases/tag/5.12)

**Comment by electrohike** [comment]
Here's the solution I've tested and applied : 

It should be possible to do quite everything from an external development environment with cross-compiling. But as I had none suitable when I tried, I did it by copying and running everything directly on the DE10-Nano board. The only thing I had to do first is extend the partition size, otherwise it was not big enough to store everything required.

Then
1. From linux-socfpga repository, download and copy all files related to version 5.12
2. From the kernel source directory, run the following commands : 
`make ARCH=arm socfpga_defconfig` otherwise when running make, the system asks for a lot - more than a hundred - of options. Even if choosing default values, it takes a very long time ...
`make oldconfig && make prepare` Those commands are shown directly in response to an attempt to make a module if the kernel has not been properly "prepared" first
`make modules` To get the necessary libraries to build modules. Otherwise any attempt still fails 
3. Edit /include/generated/utsrelease.h and set content to `#define UTS_RELEASE "5.12.0zImage"` (There's only this single line). Otherwise the modules you build won't be inserted because their version string does not mach the one of the running kernel.

And now, I can build and run my own kernel modules.
There may be other ways to do that, especially with probably less files. But at least the solution worked for may.

It is not really a mistake to correct, as I've not seen written anywhere that the Debian image is ready for kernel module development. But it may be a good idea to add a "warning" or a "comment"somewhere about that.
